File: AndroidApp/app/src/main/java/com/freya/androidapp/PermissionRequestActivity.java
package com.freya.androidapp;

import android.os.Bundle;

import androidx.activity.ComponentActivity;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;

public final class PermissionRequestActivity extends ComponentActivity {
    public static final String EXTRA_CALLBACK_PTR = "extra_callback_ptr";

    private long callbackPtr;
    private boolean resultDelivered;
    private boolean requestLaunched;

    private final ActivityResultLauncher<String> permissionLauncher =
            registerForActivityResult(
                    new ActivityResultContracts.RequestPermission(),
                    isGranted -> {
                resultDelivered = true;
                        RustLocationHelper.onPermissionResult(callbackPtr, isGranted);
                        finish();
                    }
            );

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        callbackPtr = getIntent().getLongExtra(EXTRA_CALLBACK_PTR, 0L);
        if (callbackPtr == 0L) {
            resultDelivered = true;
            finish();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (callbackPtr != 0L && !requestLaunched && !resultDelivered) {
            requestLaunched = true;
            permissionLauncher.launch(android.Manifest.permission.ACCESS_FINE_LOCATION);
        }
    }

    @Override
    protected void onDestroy() {
        if (!resultDelivered && callbackPtr != 0L) {
            resultDelivered = true;
            RustLocationHelper.onPermissionResult(callbackPtr, false);
        }
        super.onDestroy();
    }
}

File: AndroidApp/app/src/main/java/dev/drigster/taskupeatus/PermissionRequestActivity.kt
package dev.drigster.taskupeatus

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts

class PermissionRequestActivity : ComponentActivity() {
    private var callbackPtr: Long = 0L
    private var resultDelivered = false
    private var requestLaunched = false

    private val permissionLauncher: ActivityResultLauncher<String> =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
            resultDelivered = true
            RustPermissionHelper.onPermissionResult(callbackPtr, isGranted)
            finish()
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        callbackPtr = intent.getLongExtra(EXTRA_CALLBACK_PTR, 0L)
        if (callbackPtr == 0L) {
            resultDelivered = true
            finish()
        }
    }

    override fun onStart() {
        super.onStart()
        if (callbackPtr != 0L && !requestLaunched && !resultDelivered) {
            requestLaunched = true
            permissionLauncher.launch(android.Manifest.permission.ACCESS_FINE_LOCATION)
        }
    }

    override fun onDestroy() {
        if (!resultDelivered && callbackPtr != 0L) {
            resultDelivered = true
            RustPermissionHelper.onPermissionResult(callbackPtr, false)
        }
        super.onDestroy()
    }

    companion object {
        const val EXTRA_CALLBACK_PTR = "extra_callback_ptr"
    }
}
